# Análisis inicial y trata de datos

In [1]:
# set PYTHONPATH="${PYTHONPATH}:/path/to/your/project/"

In [2]:
# Importamos las librerías necesarias
import pandas as pd # type: ignore
import numpy as np # type: ignore
# from src import funciones_soporte as fs
import psycopg2 as ps # type: ignore

In [3]:
# Importamos el archivo con el que vamos a trabajar, que en este caso es un parquet. Para que funcione el read_parquet es necesario pasarle
# el parámetro ' engine = auto/pyarrow', ambos funcionan. En caso de que se use pyarrow será necesario realizar su instalaciñon previamnente en el entorno.
df_raw = pd.read_parquet("../data/reservas_hoteles.parquet", engine='auto')
df_raw.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0                    2025-03-01     2025-03-02       113           NaN   
1                    2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad  
0                              NaN          
1                              NaN          
2     Hotel Monte Verde        1.0  Madrid  
3  Hotel Brisas del Mar        5.0  Madrid  
4  Hotel Camino del Sol        1.0  Madrid

In [4]:
# Creamos una copia de los datos originales
data = df_raw.copy()
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0                    2025-03-01     2025-03-02       113           NaN   
1                    2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad  
0                              NaN          
1                              NaN          
2     Hotel Monte Verde        1.0  Madrid  
3  Hotel Brisas del Mar        5.0  Madrid  
4  Hotel Camino del Sol        1.0  Madrid

In [5]:
# función info
def info_df(dataframe):
    """
    Función que devuelve información general sobre el DatFrame que le pasemos.


    Args:
        df (DataFrame): DataFrame con información que queramos revisar

    Returns:
        DataFrame: DataFrame con información general sobre las columnas del DataFrame que se le ha pasado a la función: tipo de datos, número de
        registros, número de valores nulos, porcentaje de los valores nulos sobre el total
    """
    info_df = pd.DataFrame()
    info_df["Tipo_dato"] = dataframe.dtypes
    info_df["numero_registros"] = [dataframe[elemento].value_counts().sum() for elemento in dataframe]
    info_df["Numero_nulos"] = round(dataframe.isnull().sum())
    info_df["%_nulos"] = round((dataframe.isnull().sum()/dataframe.shape[0])*100, 2)

    return info_df

In [6]:
# Información general del dataframe
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva         object             15098             0     0.00
id_cliente         object             15098             0     0.00
nombre             object             15098             0     0.00
apellido           object             15098             0     0.00
mail               object             15098             0     0.00
competencia          bool             15098             0     0.00
fecha_reserva      object             15098             0     0.00
inicio_estancia    object             15023            75     0.50
final_estancia     object             15023            75     0.50
id_hotel            int64             15098             0     0.00
precio_noche      float64              9874          5224    34.60
nombre_hotel       object             15098             0     0.00
estrellas         float64              9926          5172    34.26
ciudad             object             15098             0     0.00

In [7]:
# Vamos a ver las columnas que tenemos en el dataframe
data.columns

Index(['id_reserva', 'id_cliente', 'nombre', 'apellido', 'mail', 'competencia',
       'fecha_reserva', 'inicio_estancia', 'final_estancia', 'id_hotel',
       'precio_noche', 'nombre_hotel', 'estrellas', 'ciudad'],
      dtype='object')

Tras realizar una overview de los datos recibidos observamos las siguientes casuísticas:
- Tenemos 14 columnas.
- Hay 3 columnas con información sobre fechas, las cuales son de tipo object. Será necesario modificar el tipo de dato de las mismas para poder trabajar más adelante.
- La columna competencia es un booleano, el cual habrá que ver con que se corresponde true y false.
- Tenemos tres columnas de id: de reserva, de cliente y de hotel. Las dos primeras con tipo object, y la última con tipo int. Cambiaremos la de id_hotel a tipo object para tenerlas las tres iguales. 
- Hay existencia de nulos en cuatro columnas:
    
    **a)** inicio_estancia: tenemos un 0,5% de nulos.
    
    **b)** fin_estancia: tenemos un 0,5% de nulos. Comprobaremos si son los mismos que la columna de inicio estancia, revisar casuística y tratar.
    
    **c)** precio_noche: hay un 34,6% de nulos, un valor significativo. Será necesario tratarlos.
    
    **d)** estrellas: hay un 34,26% de nulos, un valor significativo. Será necesario tratarlos. Asimismo el tipo de dato de esta columna es float. Podemos plantear modificarlo a tipo object, a un str. 

A continuación realizamos las modificaciones necesarias de las columnas para poder trabajar con ellas. 

## Búsqueda de duplicados completamente iguales 

In [8]:
# pd.set_option('display.max_rows', None)

In [9]:
# Vamos a revisar la existencia de duplicados completamente iguales en el df. Revisaremos si se pueden eliminar
data.duplicated().sum()
# Tenemos 98 elementos duplicados en total en el dataframe
data = data.drop_duplicates()

## Modificación columnas de fecha

In [10]:
# Creamos una función que nos modifica ek tipo de dato de las columnas que le pasemos a tipo fecha
def data_fechas(dataframe, columnas):
    for col in columnas:
        dataframe[col] = pd.to_datetime(dataframe[col])
    return dataframe

In [11]:
# modificamos las columnas que queremos a tipo fecha
data = data_fechas(data, ["fecha_reserva", "inicio_estancia", "final_estancia"])

In [12]:
# comprobamos que la modificación se ha realizado de forma correcta
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object             15000             0     0.00
id_cliente               object             15000             0     0.00
nombre                   object             15000             0     0.00
apellido                 object             15000             0     0.00
mail                     object             15000             0     0.00
competencia                bool             15000             0     0.00
fecha_reserva    datetime64[ns]              9828          5172    34.48
inicio_estancia  datetime64[ns]             14925            75     0.50
final_estancia   datetime64[ns]             14925            75     0.50
id_hotel                  int64             15000             0     0.00
precio_noche            float64              9776          5224    34.83
nombre_hotel             object             15000             0     0.00
estrellas               float64              9828          5172    34.48
ciudad                   object             15000             0     0.00

Modificación realizada con éxito.

# Columna de inicio y fin de estancia
En esta columna observamos que todas las estancias son del fin de semana del 01/03/2025 y el 02/03/2025. Podemos considerar rellenar el resto de fechas con los mismos valores. 

*Observar esto en el análisis a continuación.

In [ ]:
# Vamos a comprobar los valores únicos que tenemos en la columna de inicio de estancia.
data["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [ ]:
# Vamos a comprobar los valores únicos que tenemos en la columna de final de estancia.
data["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [ ]:
# Realizamos la modificación de los valores nulos de las columnas reemplazándolas por los valores de las fechas del fin de semana correspondientes
data["inicio_estancia"] = data["inicio_estancia"].fillna("2025-03-01")
data["final_estancia"] = data["final_estancia"].fillna("2025-03-02")

In [ ]:
# Comprobamos que la modificación se ha realizado correctamente
data["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00']
Length: 1, dtype: datetime64[ns]

In [ ]:
# Comprobamos que la modificación se ha realizado correctamente
data["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00']
Length: 1, dtype: datetime64[ns]

In [60]:
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object             15000             0     0.00
id_cliente               object             15000             0     0.00
nombre                   object             15000             0     0.00
apellido                 object             15000             0     0.00
mail                     object             15000             0     0.00
competencia                bool             15000             0     0.00
fecha_reserva    datetime64[ns]              9828          5172    34.48
inicio_estancia  datetime64[ns]             15000             0     0.00
final_estancia   datetime64[ns]             15000             0     0.00
id_hotel                 object             15000             0     0.00
precio_noche            float64              9776          5224    34.83
nombre_hotel             object             15000             0     0.00
estrellas               float64              9828          5172    34.48
ciudad                   object             15000             0     0.00
col_prueba               object             15000             0     0.00

# Modificación de los id de hotel
Como los id de hotel de los hoteles que pertenecen al grupo están repetidos, vamos a crear nuevos id asignados a cada hotel. 

In [80]:
lista_hoteles = data["nombre_hotel"].unique()[1:]

In [ ]:
def cambio_id(dataframe, lista_columna):
    lista = dataframe[columna][1].unique()
    dict_hoteles_grupo = {}
    conteo = 1
    for elemento in lista:
        dict_hoteles_grupo[elemento] = conteo
        conteo += 1
    data["columna"] = data["nombre_hotel"].map(dict_hoteles_grupo).astype("Int64")

In [90]:
data["id_hotel"] = data["nombre_hotel"].map(dict_hoteles_grupo).astype("Int64")

# Modificación de los id de hotel
Como los id de hotel de los hoteles que pertenecen al grupo están repetidos, vamos a crear nuevos id asignados a cada hotel. 

## Columna valores booleanos

In [13]:
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0           NaT      2025-03-01     2025-03-02       113           NaN   
1           NaT      2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad  
0                              NaN          
1                              NaN          
2     Hotel Monte Verde        1.0  Madrid  
3  Hotel Brisas del Mar        5.0  Madrid  
4  Hotel Camino del Sol        1.0  Madrid

In [14]:
# Al intentar filtrar la columna en sí, como es un booleano no nos devuelve nada, por lo que creamos una copia de la columna de tipo str
# a través de la cual podremos realizar el filtrado entre true y false.
data["col_prueba"] = data["competencia"].astype(str)
data.head()

id_reserva                            id_cliente  \
0  5256cc90-139b-43d2-8ec5-412495d751cf  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
1  84fd6209-bd8d-4a92-bfe6-c68ee2c49271  346f2a77-33f4-4b2d-82f3-e8501e21b28d   
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   

      nombre   apellido                          mail  competencia  \
0  Feliciana     Cantón  feliciana.cantón@example.com         True   
1   Leonardo     Hierro   leonardo.hierro@example.com         True   
2      Maite  Calatayud   maite.calatayud@example.com        False   
3      Tecla      Bonet       tecla.bonet@example.com        False   
4    Amílcar     Andrés    amílcar.andrés@example.com        False   

  fecha_reserva inicio_estancia final_estancia  id_hotel  precio_noche  \
0           NaT      2025-03-01     2025-03-02       113           NaN   
1           NaT      2025-03-01     2025-03-02       194           NaN   
2    2025-02-09      2025-03-01     2025-03-02        14        119.96   
3    2025-02-08      2025-03-01     2025-03-02        38        211.53   
4    2025-02-02      2025-03-01     2025-03-02        35        378.37   

           nombre_hotel  estrellas  ciudad col_prueba  
0                              NaN               True  
1                              NaN               True  
2     Hotel Monte Verde        1.0  Madrid      False  
3  Hotel Brisas del Mar        5.0  Madrid      False  
4  Hotel Camino del Sol        1.0  Madrid      False

`Hoteles de la competencia (col True)`

Estas columnas entendemos que se consideran como competencia del grupo, por lo que nos falta información de las mismas.

Tenemos valores nulos en este caso en las columnas de inicio y fin de estancia (23 en cada una), los cuales se corresponden en ambas columnas con los mismos clientes. Asimismo hay valores nulos en la columna de precio de noche (5172), que entendemos que se corresponde con los valores nulos de los hoteles de los cuales no tenemos estrellas (nulos en columna estrellas). Existen valores nulos en la columna de fecha de reserva, de la cual no tenemos ninguna información al respecto. 

Por otro lado no tenemos los nombres de estos hoteles, sino que están puestos en el dataframe como espacios en blanco.

Para tratar estos datos realizaremos un scrapeo de los nombres de los hoteles correspondientes, sus estrellas y el precio por noche de cada uno. Una vez hayamos obtenido la información necesaria, les asignaremos un id de hotel para poder trabajar en el dataframe.


Tenemos valores duplicados en la columna de id de cliente, será necesario eliminar estos valores. 

De estos hoteles todas las estancias se han realizado entre el 01/03/2025 y el 02/03/2025. Podemos considerar rellenar el resto de fechas con los mismos valores.

In [15]:
# Filtramos por los valores que son True
hoteles_competencia = data[data["col_prueba"] == "True"]
hoteles_competencia.head()

id_reserva  \
0   5256cc90-139b-43d2-8ec5-412495d751cf   
1   84fd6209-bd8d-4a92-bfe6-c68ee2c49271   
7   4d65096e-5f12-4ce7-9fd6-dd20e3cf5194   
9   458029ad-cc9b-4d85-8290-fec170bcc02d   
11  ac0fe86e-90a0-4fe6-a02e-6ea00e9a8621   

                              id_cliente        nombre apellido  \
0   346f2a77-33f4-4b2d-82f3-e8501e21b28d     Feliciana   Cantón   
1   346f2a77-33f4-4b2d-82f3-e8501e21b28d      Leonardo   Hierro   
7   443a7307-2183-4051-b0bb-26f12fc1f478        Judith    Rivas   
9   e1a34338-a0aa-4b0e-96de-55f884eeb074        Adolfo   Cuervo   
11  2949fb90-39ab-4613-b4c1-e238bb56faf3  María Carmen  Navarro   

                                mail  competencia fecha_reserva  \
0       feliciana.cantón@example.com         True           NaT   
1        leonardo.hierro@example.com         True           NaT   
7           judith.rivas@example.com         True           NaT   
9          adolfo.cuervo@example.com         True           NaT   
11  maría carmen.navarro@example.com         True           NaT   

   inicio_estancia final_estancia  id_hotel  precio_noche nombre_hotel  \
0       2025-03-01     2025-03-02       113           NaN                
1       2025-03-01     2025-03-02       194           NaN                
7       2025-03-01     2025-03-02       131           NaN                
9       2025-03-01     2025-03-02       114           NaN                
11      2025-03-01     2025-03-02       103           NaN                

    estrellas ciudad col_prueba  
0         NaN              True  
1         NaN              True  
7         NaN              True  
9         NaN              True  
11        NaN              True

In [16]:
info_df(hoteles_competencia)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object              5172             0     0.00
id_cliente               object              5172             0     0.00
nombre                   object              5172             0     0.00
apellido                 object              5172             0     0.00
mail                     object              5172             0     0.00
competencia                bool              5172             0     0.00
fecha_reserva    datetime64[ns]                 0          5172   100.00
inicio_estancia  datetime64[ns]              5149            23     0.44
final_estancia   datetime64[ns]              5149            23     0.44
id_hotel                  int64              5172             0     0.00
precio_noche            float64                 0          5172   100.00
nombre_hotel             object              5172             0     0.00
estrellas               float64                 0          5172   100.00
ciudad                   object              5172             0     0.00
col_prueba               object              5172             0     0.00

In [17]:
# Comprobamos si los valores nulos de las columnas de inicio y final de estancia se corresponden con los mismos clientes en ambas columnas.
# hoteles_competencia[hoteles_competencia["inicio_estancia"].isnull()] == hoteles_competencia[hoteles_competencia["final_estancia"].isnull()] 

In [ ]:
# modificar directamente con los mismos valores del finde para todos los nulos
hoteles_competencia["fecha_reserva"].unique()

<DatetimeArray>
['NaT']
Length: 1, dtype: datetime64[ns]

In [19]:
hoteles_competencia["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [20]:
hoteles_competencia["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [21]:
# Comprobamos los nombres de los hoteles de la competencia
hoteles_competencia["nombre_hotel"].unique()

array([''], dtype=object)

In [22]:
# Comprobamos los id de hotel de la competencia
hoteles_competencia["id_hotel"].unique()

array([113, 194, 131, 114, 103, 181, 128, 186, 135, 117])

In [23]:
hoteles_competencia.duplicated().sum()

np.int64(0)

In [24]:
# Revisamos duplicados
hoteles_competencia[hoteles_competencia["id_cliente"].duplicated()]

id_reserva  \
1      84fd6209-bd8d-4a92-bfe6-c68ee2c49271   
1333   35888670-8393-4a86-b10f-42d8cc0cc558   
2467   20c86be7-6537-4d1e-8553-3e95b49af0b9   
3777   beda58ca-7f20-488c-b680-b6d748f666c5   
3944   3e9c3923-4b34-4566-ab96-74237b4f4eae   
4826   fbba658c-f025-484c-a3a4-7a7a8af42863   
8373   8b3a1b34-68dd-4429-93fa-c561ca8137c6   
8551   dcfaed8a-ce16-47f6-8eae-69c410a4ea8b   
11055  9dcb96fc-ac04-4d47-b7e3-b19705c0b5bd   
11858  9387407f-a17f-4ccb-b0b9-fc8e77b961ed   
14625  71bfa845-b10b-42b3-9e09-fcb238341936   

                                 id_cliente            nombre    apellido  \
1      346f2a77-33f4-4b2d-82f3-e8501e21b28d          Leonardo      Hierro   
1333   bfbb9754-fcab-4580-bb2a-9ab295149227           Teodora        Seco   
2467   37ec97b1-9ceb-48e0-a3af-7a79b9585603               Ruy        Ríos   
3777   2b15a636-76f1-4c9a-b689-0581c3245aa2               Ana  Valenzuela   
3944   a44bb598-0a86-4e4a-b19e-66ddff9f3d2f           Manuela     Fonseca   
4826   21294898-0afc-489d-8f71-1ca78bedbe2f              Lina       Ariño   
8373   5a2d9076-03e5-482c-a7a4-45f5abc6cc62            Álvaro     Pascual   
8551   f3e84e3b-35a9-4207-a6a1-4cff7487f569          Consuelo     Estévez   
11055  f8fc8162-f481-44f2-ac98-4a952cbe8ffc  Francisco Javier       Sainz   
11858  9e659fbd-bcc6-4236-9787-6bca04d567c1           Ciríaco       Viana   
14625  1ae69965-f7b0-46d1-ab3a-0b6d207db369              Chus     Uriarte   

                                     mail  competencia fecha_reserva  \
1             leonardo.hierro@example.com         True           NaT   
1333             teodora.seco@example.com         True           NaT   
2467                 ruy.ríos@example.com         True           NaT   
3777           ana.valenzuela@example.com         True           NaT   
3944          manuela.fonseca@example.com         True           NaT   
4826               lina.ariño@example.com         True           NaT   
8373           álvaro.pascual@example.com         True           NaT   
8551         consuelo.estévez@example.com         True           NaT   
11055  francisco javier.sainz@example.com         True           NaT   
11858           ciríaco.viana@example.com         True           NaT   
14625            chus.uriarte@example.com         True           NaT   

      inicio_estancia final_estancia  id_hotel  precio_noche nombre_hotel  \
1          2025-03-01     2025-03-02       194           NaN                
1333       2025-03-01     2025-03-02       128           NaN                
2467       2025-03-01     2025-03-02       131           NaN                
3777       2025-03-01     2025-03-02       103           NaN                
3944       2025-03-01     2025-03-02       117           NaN                
4826       2025-03-01     2025-03-02       194           NaN                
8373       2025-03-01     2025-03-02       103           NaN                
8551       2025-03-01     2025-03-02       131           NaN                
11055      2025-03-01     2025-03-02       114           NaN                
11858      2025-03-01     2025-03-02       181           NaN                
14625      2025-03-01     2025-03-02       181           NaN                

       estrellas ciudad col_prueba  
1            NaN              True  
1333         NaN              True  
2467         NaN              True  
3777         NaN              True  
3944         NaN              True  
4826         NaN              True  
8373         NaN              True  
8551         NaN              True  
11055        NaN              True  
11858        NaN              True  
14625        NaN              True

In [25]:
# hoteles_competencia

`Hoteles del grupo (col False)`

Estas columnas entendemos que no se consideran como competencia del grupo, por lo que tenemos casi toda la información de las mismas. Tenemos los nombres de los hoteles del grupo, junto a sus id correspondientes.

Tenemos valores nulos en este caso en las columnas de inicio, fin de estancia y precio de noche (52 en cada una), los cuales se corresponden en todas las columnas con los mismos clientes.

Tenemos duplicados en las columnas de id de cliente (178) e id de reserva (98)

In [26]:
# Filtramos por las columnas que contienen False.
hoteles_grupo = data[data["col_prueba"] == "False"]
hoteles_grupo.head()

id_reserva                            id_cliente  \
2  40c4cb55-d1f5-407b-832f-4756b8ff77b4  6ec4d7d2-e382-47cc-94f6-c647577d4035   
3  f2ce8df5-7844-43e1-8c0f-97ce7a208f21  aed17fe8-eb18-4f69-bf08-df1afdc339c3   
4  57d4515a-447d-4067-afcb-9bdcf4e4e915  710a2fbb-264f-426a-aa66-f8fffd146a2d   
5  ccbd9fe9-5a60-4f6f-bed3-73a7158ecba7  df42ae3d-0b9b-4c06-b22e-939323833ef8   
6  99c3dc4f-663c-45f7-849a-ac9313f3746a  a16a9d83-6888-4947-a7ab-ed41203e347c   

    nombre   apellido                         mail  competencia fecha_reserva  \
2    Maite  Calatayud  maite.calatayud@example.com        False    2025-02-09   
3    Tecla      Bonet      tecla.bonet@example.com        False    2025-02-08   
4  Amílcar     Andrés   amílcar.andrés@example.com        False    2025-02-02   
5     Joan    Vazquez     joan.vazquez@example.com        False    2025-02-08   
6    Chelo       Flor       chelo.flor@example.com        False    2025-02-09   

  inicio_estancia final_estancia  id_hotel  precio_noche  \
2      2025-03-01     2025-03-02        14        119.96   
3      2025-03-01     2025-03-02        38        211.53   
4      2025-03-01     2025-03-02        35        378.37   
5      2025-03-01     2025-03-02        36         85.45   
6      2025-03-01     2025-03-02        43        309.80   

             nombre_hotel  estrellas  ciudad col_prueba  
2       Hotel Monte Verde        1.0  Madrid      False  
3    Hotel Brisas del Mar        5.0  Madrid      False  
4    Hotel Camino del Sol        1.0  Madrid      False  
5  Hotel Puerta del Cielo        5.0  Madrid      False  
6      Hotel Encanto Real        1.0  Madrid      False

In [27]:
info_df(hoteles_grupo)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object              9828             0     0.00
id_cliente               object              9828             0     0.00
nombre                   object              9828             0     0.00
apellido                 object              9828             0     0.00
mail                     object              9828             0     0.00
competencia                bool              9828             0     0.00
fecha_reserva    datetime64[ns]              9828             0     0.00
inicio_estancia  datetime64[ns]              9776            52     0.53
final_estancia   datetime64[ns]              9776            52     0.53
id_hotel                  int64              9828             0     0.00
precio_noche            float64              9776            52     0.53
nombre_hotel             object              9828             0     0.00
estrellas               float64              9828             0     0.00
ciudad                   object              9828             0     0.00
col_prueba               object              9828             0     0.00

In [28]:
# Comprobamos si los valores nulos de las columnas de inicio y final de estancia se corresponden con los mismos clientes en ambas columnas.
# df_false[df_false["inicio_estancia"].isnull()] == df_false[df_false["final_estancia"].isnull()]

In [29]:
hoteles_grupo["inicio_estancia"].unique()

<DatetimeArray>
['2025-03-01 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [30]:
hoteles_grupo["final_estancia"].unique()

<DatetimeArray>
['2025-03-02 00:00:00', 'NaT']
Length: 2, dtype: datetime64[ns]

In [31]:
# Vamos a revisar los nombres de los hoteles y su id_correspondiente
hoteles_grupo[["id_reserva","id_hotel","nombre_hotel","estrellas"]].sort_values("id_hotel")

id_reserva  id_hotel            nombre_hotel  \
69     ebd89c70-66d4-41d1-b218-dc9113ce6ce6         2      Hotel Vista Alegre   
14944  ce0f0e78-79ef-4662-bf46-ac07d42cf944         2  Hotel Puerta del Cielo   
36     728efa5a-8a5a-4e53-8b2b-349c4fcab5d9         2     Hotel Luz de Madrid   
12769  df75281f-48e6-4913-b671-18368d3ee071         2    Hotel Camino del Sol   
6398   f08d06e4-ecf6-483b-a8e5-2f227dc69903         2      Hotel Encanto Real   
...                                     ...       ...                     ...   
1828   01f8dd25-f627-4cec-a6e8-ae5fce39e63a        49        Hotel Sol y Luna   
1825   eddb605a-abca-4ee0-9037-99b55fab8ab7        49      Hotel Vista Alegre   
1817   7b31ede3-7638-4080-a127-52bab7a07cfa        49     Hotel Los Almendros   
13372  f2b7d712-c4b7-49d0-ae26-de620cb77840        49  Hotel Jardines del Rey   
7950   553e1674-134b-43b3-85a2-8e73e460f4a0        49     Hotel Rincón Sereno   

       estrellas  
69           3.0  
14944        5.0  
36           3.0  
12769        2.0  
6398         4.0  
...          ...  
1828         2.0  
1825         2.0  
1817         4.0  
13372        4.0  
7950         3.0  

[9828 rows x 4 columns]

In [32]:
hoteles_grupo.duplicated().sum()

np.int64(0)

In [33]:
# Revisamos duplicados de los valores 
hoteles_grupo["id_reserva"].duplicated().sum()

np.int64(0)

In [34]:
hoteles_grupo["id_cliente"].duplicated().sum()

np.int64(80)

In [35]:
hoteles_grupo[hoteles_grupo["mail"].duplicated()]

id_reserva  \
3007   904807cd-f9de-4261-b2ef-905c818e5c28   
3089   5223c253-9e19-4fe7-bace-56e2d8f97141   
3112   868a22d6-890a-4533-9db7-d5f2317e0598   
4373   1302272f-354c-4f54-acc0-b1a42f9ad9be   
4545   c461cfa8-1732-45b8-b6d8-1ec110f6a7da   
4646   2c20e5d9-c7b9-4d33-912c-2e06b6c38351   
5604   77c97c1a-e649-4590-9537-6e7e976e8fd6   
5623   e9b32558-762e-4986-93b9-69198c7e1b5f   
5732   82b22e29-2068-484e-9df5-9a0609ea4990   
6492   f5e75c3f-26a1-4f90-b69d-793ba6bd911e   
7249   4b51d50c-7e3d-4157-a118-944f1488eedb   
7313   9d84ec11-9d3a-4adf-aa61-04465f4d271a   
7337   5a1077b5-ce3d-4f59-876c-7a2263285533   
8016   994040b7-14d5-4630-8c33-f1871d566878   
8744   13483be5-a8ec-434b-9eb3-492815565929   
8973   78787e6c-818f-4a91-b803-7b12ba9bd6f7   
9577   8185208b-6e0a-4685-9adb-a2c238be6a57   
10148  129dd38a-eb56-4898-b5d9-e0d81605e838   
10247  328e1254-1f4b-4788-9967-9d88cbdfd723   
10343  8dcc242d-c00f-4c5d-ab06-952da2db147f   
10348  e35f83da-6935-486b-aeb9-e084a3685906   
10818  31d63bfc-5103-414c-8325-38d0404eac90   
11140  08c77941-c394-45fc-bef2-c45ee3fa63ac   
11429  7b52ddeb-efa1-4d8e-aeca-e81bb41d2596   
12406  bcd8c15f-6190-497a-9031-873f9f34c950   
12669  e03aea99-3d55-47f7-ae77-57b054f261c3   
12711  c6152f53-20e1-4705-9a81-a749777bb697   
12985  1c827007-789d-45e9-b365-4799b0e569df   
13146  305ee50c-ff72-4fb8-88fa-b1a879e66852   
13263  5533ecfa-d3e8-44d6-b648-094dcc2bbbeb   
13395  4f9adeeb-fabe-4748-a3b7-3bffa9cae5dd   
13429  2736ef8b-c5a5-4e9a-b048-269ec5400640   
13493  814861d9-1e44-43e2-a176-100575a1c048   
13501  f6ee85c3-8026-41ad-95b9-ccacd4c0f3e0   
13549  8b0d64b3-9671-4c98-9871-5138eb0ae909   
13573  76858380-64f2-4fe2-b8e8-8d416e138285   
13696  096e97db-65d5-4223-a11a-2f30adf4467b   
13755  2068e77f-4383-4a33-9aed-8055421e3b79   
14133  7b100c76-cfe5-47fe-8317-559449a9cbc9   
14213  28f614a4-f2c5-4079-9a82-4942426a6f6c   
14263  ebcaa11a-b1fd-49e7-8b4d-9190c72bd578   
14845  b9c1b90a-a161-4837-9298-21d4fc6f68eb   
14909  8b9e296b-dcba-4c82-8e0a-91b3d051d2ce   

                                 id_cliente      nombre    apellido  \
3007   1d97512a-8d30-43a2-9ee4-043b4107a1b1    Ceferino        Sosa   
3089   7c72ec8f-aff0-4e75-a8a0-75dcbdf66c13      Sergio     Carrión   
3112   de536f30-cedd-4991-a354-4e1cea133e53   Adalberto     Atienza   
4373   7c093268-75f6-49d5-bd5f-1b585348b1e5   Manuelita       Leiva   
4545   9920620c-67c7-4263-a513-c709e37a7f43      Felisa    Benavent   
4646   db8a4c3d-557d-416a-99fc-3bc618694a8b    Consuela       Folch   
5604   8d6533d3-2c10-41a9-b400-42c5bec2e048    Graciana     Cánovas   
5623   fe6bf3eb-667e-460e-83b2-1b0ef4d1c2da      Olalla     Morillo   
5732   f52550af-fd0a-4724-ac87-a1ad164282a2     Leandra   Castañeda   
6492   a7ddcd0c-a31b-4d46-a1e9-62b8221f35ad     Ariadna      Miguel   
7249   0693f9ba-1a85-41d5-9e1a-4756cb4a2905        Maxi      Ibañez   
7313   1a8c6a88-38fc-4fcc-8279-bf867859dd0a   Saturnina     Moliner   
7337   ba3d075a-f661-41e4-827d-8de81d922ea3       Celso      Elorza   
8016   c7bb2113-172a-463c-b61b-06e6527a1ad4     Edgardo      Gibert   
8744   8abb9d81-383e-4318-aeb7-6ff4c62ea7b4     Alberto  Montserrat   
8973   3852feaf-2c81-436e-aa77-95ef2f28cf94    Emiliano     Diéguez   
9577   5440e285-2da0-495d-ae35-e83bdaf29f28     Abigaíl       Ayala   
10148  997e92ff-707e-425d-8bf2-05884ec58d2a    Epifanio      Rincón   
10247  fa5a8b4d-fc2a-4dba-9d65-36869de8fed9     Lorenza      Aguiló   
10343  d966049b-f451-469f-a60a-fd345938df06      Noelia    Higueras   
10348  889fd392-f8a3-4bd1-86c0-da82571b4843      Samuel        Arco   
10818  901a6420-3ed6-410f-bc34-1dcc889bdffb   Esperanza    Gonzalez   
11140  d65b4528-3c1d-4fc0-97b7-602ba3a9b023        Eric   Salamanca   
11429  210636b2-8d32-48df-8da0-cf80f981e0a2   Eleuterio   Castrillo   
12406  c8b8e728-f11f-4001-ba70-f11ca5b20cce  Victoriano     Ródenas   
12669  a2d2c4ee-e64b-41e4-abef-44676dc768f2     Fabiola     Monreal   
12711  6ea53a8c-f452-4ffe-a451-e0ffce86d68a    

In [36]:
hoteles_grupo[hoteles_grupo["id_cliente"].duplicated()]

id_reserva  \
557    35c7c9d8-4898-4d05-9c34-099e71348a64   
583    172dc5c0-ed32-4f1c-ac40-21f25d3e0742   
592    d75f9705-4c2a-4f9f-a080-584a94eb72b5   
653    47b96396-8992-40d5-a15f-0ad4c74ca2c2   
686    5b3f7171-ea11-4cd5-9e8f-037085d190c0   
...                                     ...   
13923  e63ee24d-eadd-40d9-b4ab-ef898ad6d6ce   
14259  76f0c15e-8ac9-4dfe-8e89-6d575d3bd633   
14743  98041532-2e84-42e8-85fd-26a7cba77ce4   
14770  02da8794-d87e-4970-8d6a-624f2e0df1a4   
14821  332ce6a5-c3b8-4af4-bd45-7f9ed4855c04   

                                 id_cliente    nombre    apellido  \
557    e13ac13a-2b7f-4fb6-a98f-5cd417859a23  Adelardo      Bernad   
583    cba30122-d7aa-4bca-8b57-70eb0007ff95    Martín    Villalba   
592    614c4625-81ae-4cb8-be62-beef3ba967bb    Nayara       Vives   
653    f5cf6d6a-eec3-465a-a5da-f8ac77410347  Marciano       Tomas   
686    e444732b-9473-4bef-b72f-fe03459e4f01     Leyre     Mínguez   
...                                     ...       ...         ...   
13923  9920620c-67c7-4263-a513-c709e37a7f43   Teodoro       Gomez   
14259  46b9766f-eb3a-47c5-9eff-6a77bfacc6a5    Eliana     Marquez   
14743  4611be7e-ebe0-4ea9-904b-970f651bf488    Wálter    Corbacho   
14770  384d82f7-c1aa-4c61-b0c6-9e02606d67da      Adán  Montenegro   
14821  f4d72083-c218-4e30-b4c2-e1d0608ceeca     Charo    Torrijos   

                              mail  competencia fecha_reserva inicio_estancia  \
557    adelardo.bernad@example.com        False    2025-02-01      2025-03-01   
583    martín.villalba@example.com        False    2025-02-05      2025-03-01   
592       nayara.vives@example.com        False    2025-02-04      2025-03-01   
653     marciano.tomas@example.com        False    2025-02-03      2025-03-01   
686      leyre.mínguez@example.com        False    2025-02-10      2025-03-01   
...                            ...          ...           ...             ...   
13923    teodoro.gomez@example.com        False    2025-02-01      2025-03-01   
14259   eliana.marquez@example.com        False    2025-02-03      2025-03-01   
14743  wálter.corbacho@example.com        False    2025-02-11      2025-03-01   
14770  adán.montenegro@example.com        False    2025-02-11      2025-03-01   
14821   charo.torrijos@example.com        False    2025-02-10      2025-03-01   

      final_estancia  id_hotel  precio_noche            nombre_hotel  \
557       2025-03-02        33        178.04    Hotel Camino del Sol   
583       2025-03-02        19         82.85     Hotel Los Almendros   
592       2025-03-02        15        258.34    Hotel Camino del Sol   
653       2025-03-02         6        274.93     Hotel Los Almendros   
686       2025-03-02        33        428.72     Hotel Luz de Madrid   
...              ...       ...           ...                     ...   
13923     2025-03-02        38        375.39        Hotel Costa Azul   
14259     2025-03-02        36         90.53  Hotel Jardines del Rey   
14743     2025-03-02        40        180.28  Hotel Jardines del Rey   
14770     2025-03-02         6        140.71      Hotel Encanto Real   
14821     2025-03-02        28        332.32        Hotel Costa Azul   

       estrellas  ciudad col_prueba  
557          5.0  Madrid      False  
583          3.0  Madrid      False  
592          4.0  Madrid      False  
653          3.0  Madrid      False  
686          3.0  Madrid      False  
...          ...     ...        ...  
13923        4.0  Madrid      False  
14259        5.0  Madrid      False  
14743        3.0  Madrid      False  
14770        4.0  Madrid      False  
14821        2.0  Madrid      False  

[80 rows x 15 columns]

## Columna id_hotel

In [37]:
# Tenemos el id del hotel como tipo int, pero nos interesa modificarlo a tipo str/object
data["id_hotel"].dtypes

dtype('int64')

In [38]:
# Realizamos la modificación del tipo de dato
data["id_hotel"] = data["id_hotel"].astype(str)

In [39]:
# Comprobamos que se ha realizado la modificación correctamente
info_df(data)

Tipo_dato  numero_registros  Numero_nulos  %_nulos
id_reserva               object             15000             0     0.00
id_cliente               object             15000             0     0.00
nombre                   object             15000             0     0.00
apellido                 object             15000             0     0.00
mail                     object             15000             0     0.00
competencia                bool             15000             0     0.00
fecha_reserva    datetime64[ns]              9828          5172    34.48
inicio_estancia  datetime64[ns]             14925            75     0.50
final_estancia   datetime64[ns]             14925            75     0.50
id_hotel                 object             15000             0     0.00
precio_noche            float64              9776          5224    34.83
nombre_hotel             object             15000             0     0.00
estrellas               float64              9828          5172    34.48
ciudad                   object             15000             0     0.00
col_prueba               object             15000             0     0.00

In [40]:
arch_limpio = data.to_pickle("../data/reservas_hoteles_limpio.pkl")
arch_limpio